src/openforms/submissions/tests/test_post_submission_event.py

```diff
@@ -0,0 +1,930 @@
+        )
+
+        with patch(
+            "openforms.registrations.contrib.email.plugin.EmailRegistration.register_submission"
+        ) as mock_registration, patch(
+            "openforms.registrations.contrib.email.plugin.EmailRegistration.update_payment_status"
+        ) as mock_payment_status_update:
+            with self.captureOnCommitCallbacks(execute=True):
+                on_post_submission_event(submission.id)
+
+        submission.refresh_from_db()
+
+        self.assertTrue(SubmissionReport.objects.filter(submission=submission).exists())
+        mock_registration.assert_not_called()
+        mock_payment_status_update.assert_not_called()
+
+        mails = mail.outbox
+
+        self.assertEqual(1, len(mails))
+        self.assertEqual(
+            mails[0].subject, "Confirmation of your Pretty Form submission"
+        )
+        self.assertEqual(mails[0].to, ["test@test.nl"])
+        self.assertEqual(mails[0].cc, [])
+
+        cosign_info = "This email is a confirmation that this form has been co-signed by cosign@test.nl and can now be processed."
+        payment_info = (
+            "Payment of EUR 10.00 is required. You can pay using the link below."
+        )
+
+        self.assertNotIn(cosign_info, mails[0].body.strip("\n"))
+        self.assertNotIn(payment_info, mails[0].body.strip("\n"))
+
+        submission.refresh_from_db()
+
+        self.assertFalse(submission.cosign_confirmation_email_sent)
+        self.assertTrue(submission.payment_complete_confirmation_email_sent)
+        self.assertEqual(submission.auth_info.value, "111222333")
+
+    def test_retry_flow(self):
+        submission = SubmissionFactory.from_components(
+            components_list=[
+                {
+                    "key": "email",
+                    "type": "email",
+                    "label": "Email",
+                    "confirmationRecipient": True,
+                },
+                {
+                    "key": "cosign",
+                    "type": "cosign",
+                    "label": "Cosign component",
+                    "validate": {"required": True},
+                },
+            ],
+            submitted_data={"email": "test@test.nl", "cosign": "cosign@test.nl"},
+            completed=True,
+            cosign_request_email_sent=True,
+            cosign_complete=True,
+            cosign_confirmation_email_sent=True,
+            payment_complete_confirmation_email_sent=True,
+            confirmation_email_sent=True,
+            form__registration_backend="email",
+            form__registration_backend_options={"to_emails": ["test@registration.nl"]},
+            form__name="Pretty Form",
+            auth_info__attribute=AuthAttribute.bsn,
+            auth_info__value="111222333",
+            form__product__price=10,
+            form__payment_backend="demo",
+            needs_on_completion_retry=True,
+            registration_status=RegistrationStatuses.failed,
```

Comment:
is there a reason you don't use the traits like `registration_failed=True`? They generally help enforce business logic state, like ensuring that `completed=True` and a `last_register_date` is set.